### PR TITLE
bg/lwpmiopen 193 : Integrate CK's batch norm backward training into non-tunable MIOpen solver

### PR DESCRIPTION
**Pull request by bghimireamd**
* enable BF16 in `miopenBatchNormalizationBackward`
* add CK's Batch Norm Backwards solver for `F16,  F32, F64, BF16` types
* NHWC layout
* move batch norm NHWC (test/bn_spatial_nhwc_test.cpp) to bn.hpp

### CODE REVIEW

**Comment by junliume** [comment]
@bghimireamd we are having some problems with the test:
```
[2023-09-12T09:05:07.776Z] /home/jenkins/workspace/_MIOpen_bg_LWPMIOPEN-193_bn_back/build/bin/test_bn_spatial_nhwc_test --half --input {4, 64, 28, 28}, {50176, 1, 1792, 64}, packed,  
[2023-09-12T09:05:07.776Z] FAILED: -nan
[2023-09-12T09:05:07.776Z] Iteration: 0
[2023-09-12T09:05:07.776Z] Backward Batch Spatial Normalization Recalc Mean and Variance: 
[2023-09-12T09:05:07.776Z] X Input tensor: 4, 64, 28, 28
[2023-09-12T09:05:07.776Z] Delta Y Input tensor: 4, 64, 28, 28
[2023-09-12T09:05:07.776Z] Delta X output tensor output failed verification.
[2023-09-12T09:05:07.776Z] Max diff: 47584
[2023-09-12T09:05:07.776Z] Mismatch at 0: 0.000249147 != 6.37891
[2023-09-12T09:05:07.776Z] Non finite number found in gpu at 18: -nan
[2023-09-12T09:05:07.776Z] /home/jenkins/workspace/_MIOpen_bg_LWPMIOPEN-193_bn_back/build/bin/test_bn_spatial_nhwc_test --half --input {4, 64, 28, 28}, {50176, 1, 1792, 64}, packed,  
[2023-09-12T09:05:07.776Z] FAILED: -nan
[2023-09-12T09:05:07.776Z] Iteration: 0
[2023-09-12T09:05:07.776Z] Backward Batch Spatial Normalization Use Saved Mean and Variance: 
[2023-09-12T09:05:07.776Z] X Input tensor: 4, 64, 28, 28
[2023-09-12T09:05:07.776Z] Delta Y Input tensor: 4, 64, 28, 28
[2023-09-12T09:05:07.776Z] Delta X output tensor output failed verification.
[2023-09-12T09:05:07.776Z] Max diff: 42560
[2023-09-12T09:05:07.776Z] Mismatch at 0: 0.000249147 != 0.000611782
[2023-09-12T09:05:07.776Z] Non finite number found in gpu at 106: -nan

File: test/fusionHost.hpp [hunk 1]
```diff
@@ -280,14 +280,14 @@ void batchNormSpatialHostFwdTrain(const tensor<T>& input,
 }
 
 template <class T, class U>
-void batchNormSpatialHostBwdTrain(const tensor<T>& x_input,
+void batchNormSpatialHostBwdTrain(const tensor<U>& x_input,
```
Comment (CAHEK7) [inline test/fusionHost.hpp]: I know it's an old code and you've just rearranged the types. But may I ask to rename the types as well? It's not so clear what does `U` and `T` mean. And why some types were replaced.

**Comment by CAHEK7** [comment]
Could you please apply #2379

**Comment by junliume** [comment]
> Could you please apply #2379

Ping @bghimireamd :)

File: test/gtest/bn_test_data.hpp [hunk 2]
```diff
@@ -226,3 +227,115 @@ struct BNInferTestData : public BNTestData<XDataType, YDataType, TConfig>
...
+    }
+
+    void InitTensorsWithRandValue()
+    {
+        std::random_device rd{};
+        std::mt19937 gen{rd()};
+        std::uniform_int_distribution<> d{0, 100};
+        auto gen_value = [&](auto...) {
+            return 1e-2 * static_cast<ScaleDataType>(d(gen)) * ((d(gen) % 2 == 1) ? -1 : 1);
+        };
```
Comment (CAHEK7) [inline test/gtest/bn_test_data.hpp]:
```suggestion
        auto gen_value = [](auto...) {
            return prng::gen_descreet_uniform_sign(static_cast<ScaleDataType>(1e-2), 100);
        };
```
See #2397 for details.

Comment (CAHEK7) [reply test/gtest/bn_test_data.hpp]: Also `#include <random>` at line 28 should be replaced with `#include "random.hpp"`.

**Review by CAHEK7 — CHANGES_REQUESTED** [review]

File: test/gtest/bn_test_data.hpp [hunk 3]
```diff
@@ -226,3 +227,115 @@ struct BNInferTestData : public BNTestData<XDataType, YDataType, TConfig>
...
+        dy.generate(gen_value);
+        bnScale.generate(gen_value);
+        savedMean.generate(gen_value);
+
+        auto gen_var = [&](auto...) { return 1e-2 * (static_cast<MeanVarDataType>(d(gen)) + 1); };
```
Comment (CAHEK7) [inline test/gtest/bn_test_data.hpp]:
```suggestion
        auto gen_var = [](auto...) {
            return static_cast<T>(1e-2) * static_cast<MeanVarDataType>(prng::gen_0_to_B(100) + 1);
        };
```

**Review by CAHEK7 — DISMISSED** [review]

**Review by junliume — APPROVED** [review]
